docs: payee registration and fee distribution for relayer operators (#1577)
cosmos/ibc-go: docs/middleware/ics29-fee/fee-distribution.md

<!--
order: 4
-->

> # Fee Distribution

Reviewer: nit: trying to follow [Google documentation style guide](https://developers.google.com/style/headings) (_Use sentence case for headings and titles._). I promise I will also add to the repo a documentation style guide to document all these little nits.

```suggestion
# Fee distribution
```

> Learn about payee registration for distribution of packet fees. The following document is intended for relayer operators {synopsis}

Reviewer 2: ```suggestion
Learn about payee registration for the distribution of packet fees. The following document is intended for relayer operators. {synopsis}
```

> Learn about payee registration for the distribution of packet fees. The following document is intended for relayer operators. {synopsis}

Reviewer 3: Nice. This is a meta-level comment but perhaps we can have who the document is intended for as its own seperate section at the top?

> ## Pre-requisite Readings

Reviewer: ```suggestion
## Pre-requisite readings
```

* [Fee Middleware](overview.md) {prereq}

Packet fees are divided into 3 distinct amounts in order to compensate relayer operators for packet relaying on fee enabled IBC channels.

> - `RecvFee`: The sum of all packet receive fees distributed to a payee for successful execution of `MsgRecvPacket`.

Reviewer 2: Seeing as we're saying the `sum of all packet receive fees` it might be worth noting above that a packet can be `topped` up multiple times. wdyt?

Reviewer: Maybe that could be added to the section [Escrowing and paying out fees](https://github.com/cosmos/ibc-go/blob/21ac60f50040ffbef6b34e7a8ae3061d4749b224/docs/middleware/ics29-fee/msgs.md#escrowing-and-paying-out-fees) of @charleenfei's PR.

Author: Yeah, possibly adding to @charleenfei's PR would be a nice idea and segway into linking to this docs page.

https://github.com/cosmos/ibc-go/pull/1572#discussion_r905222245

- `AckFee`: The sum of all packet acknowledgement fees distributed to a payee for successful execution of `MsgAcknowledgement`.
- `TimeoutFee`: The sum of all packet timeout fees distributed to a payee for successful execution of `MsgTimeout`.

## Register a payee address for forward relaying

> As mentioned previously in [ICS29 Concepts](../ics29-fee/overview.md#concepts), the forward relayer describes the actor who performs the submission of `MsgRecvPacket` on the destination chain. 

Reviewer: I think I would remove "previously" since nobody guarantees that the reader will have read the other section before...

```suggestion
As mentioned in [ICS29 Concepts](../ics29-fee/overview.md#concepts), the forward relayer describes the actor who performs the submission of `MsgRecvPacket` on the destination chain. 
```

Fee distribution for incentivized packet relays takes place on the packet source chain. 

> Relayer operators are expected to register a counterparty payee address, in order to be compensated accordingly with `RecvFees` upon completion of a packet lifecycle. 

Reviewer: ```suggestion
Relayer operators are expected to register a counterparty payee address, in order to be compensated accordingly with `RecvFee`s upon completion of a packet lifecycle. 
```

The counterparty payee address registered on the destination chain is encoded into the packet acknowledgement and communicated as such to the source chain for fee distribution.
If a counterparty payee is not registered for the forward relayer on the destination chain, the escrowed fees will be refunded upon fee distribution.

> A transaction must be submitted to the desintation chain including a `CounterpartyPayee` address of an account on the source chain.

Reviewer 3: destination

The transaction must be signed by the `Relayer`.

```go
type MsgRegisterCounterpartyPayee struct {
	// unique port identifier
	PortId string
	// unique channel identifier
	ChannelId string
	// the relayer address
	Relayer string
	// the counterparty payee address
	CounterpartyPayee string
}
```

This message is expected to fail if:

- `PortId` is invalid (see [24-host naming requirements](https://github.com/cosmos/ibc/blob/master/spec/core/ics-024-host-requirements/README.md#paths-identifiers-separators).
- `ChannelId` is invalid (see [24-host naming requirements](https://github.com/cosmos/ibc/blob/master/spec/core/ics-024-host-requirements/README.md#paths-identifiers-separators)).
- `Relayer` is an invalid address (see [Cosmos SDK Addresses](https://github.com/cosmos/cosmos-sdk/blob/main/docs/basics/accounts.md#Addresses)).
- `CounterpartyPayee` is empty.

See below for an example CLI command:

```
simd tx ibc-fee register-counterparty-payee transfer channel-0 cosmos1rsp837a4kvtgp2m4uqzdge0zzu6efqgucm0qdh osmo1v5y0tz01llxzf4c2afml8s3awue0ymju22wxx2 --from cosmos1rsp837a4kvtgp2m4uqzdge0zzu6efqgucm0qdh
```

## Register a payee address for reverse and timeout relaying

> As mentioned previously in [ICS29 Concepts](../ics29-fee/overview.md#concepts), the reverse relayer describes the actor who performs the submission of `MsgAcknowledgement` on the source chain. 

Reviewer: ```suggestion
As mentioned in [ICS29 Concepts](../ics29-fee/overview.md#concepts), the reverse relayer describes the actor who performs the submission of `MsgAcknowledgement` on the source chain. 
```

> Similarly the timeout relayer describes the actor who performs the submission of `MsgTimeout` on the source chain.

Reviewer: ```suggestion
Similarly the timeout relayer describes the actor who performs the submission of `MsgTimeout` (or `MsgTimeoutOnClose`) on the source chain.
```

> Relayer operators may choose to register an optional payee address, in order to be compensated accordingly with `AckFees` and `TimeoutFees` upon completion of a packet life cycle.

Reviewer: ```suggestion
Relayer operators may choose to register an optional payee address, in order to be compensated accordingly with `AckFee`s and `TimeoutFee`s upon completion of a packet life cycle.
```

> If a payee is not registered for the reverse or timeout relayer on the source chain, then fee distribution assumes the default behaviour, where fees are paid out to the relayer account which delivers `MsgAcknowledgement` or `MsgTimeout`.

Reviewer: ```suggestion
If a payee is not registered for the reverse or timeout relayer on the source chain, then fee distribution assumes the default behaviour, where fees are paid out to the relayer account which delivers `MsgAcknowledgement` or `MsgTimeout`/`MsgTimeoutOnClose`.
```
